File: hw2/p1/p1.cpp
// NAME       : Aaron Graham
// COURSE     : NERS 544
// ASSIGNMENT : Homework 2
// PROBLEM    : Problem 1
// DATE       : February 23, 2015

#define PI 3.14159265358979323
#define SIGMA 0.35
#define C 1.05078216355812
//#define NORM 1.029149923805585
//#define NORMX2 0.1211987601379424

#include<iostream>
#include<cstdlib>
#include<time.h>
#include<cmath>

using namespace std;

double normRand(void)
{
  return static_cast<double>(rand())/static_cast<double>(RAND_MAX);
}

double sample(void)
{
  return SIGMA*sqrt(-2.0*log(normRand()))*cos(2.0*PI*normRand());
}

double g(double x)
{
  return exp(-(x*x)/(2.0*SIGMA*SIGMA))/(SIGMA*sqrt(2.0*PI));
}

double f(double x)
{
  double tmp = cos(PI*x/2.0);
  return (3.0*PI/8.0)*tmp*tmp*tmp;
}

int main()
{
  srand(time(NULL));

  double x, y, test, bound, Ex2 = 0.0;
  int N = 1e6, i;
  bool picking = false;

  for (i = 0; i < N; i++)
  {
    picking = true;
    while (picking)
    {
      x = sample();
      y = f(x);
      bound = g(x);
      test = normRand()*C*bound;
      if (test <= y) picking = false;
    }
    Ex2 += x*x;
  }

  Ex2 /= static_cast<double>(N);

  cout << "After " << N << " iterations, E[X^2]=" << Ex2 << "." << endl;

  return 0;

}

File: hw2/p1/p1.m
close all, clear all, clc
format long;

sigma = 0.35;
c = 1.05078216355812;

N = 1e6;
Ex = 0;
Ex2 = 0;
samples = zeros(2,N);

for i=1:N
  picking = 1;
  while (picking)
%     if (rand() > 0.5)
%       x=sqrt(-2*sigma*sigma*log(rand()));
%     else
%       x=-sqrt(-2*sigma*sigma*log(rand()));
%     end
    x = sigma*sqrt(-2*log(rand()))*cos(2*pi*rand());
    tmp = cos(pi*x/2);
    y = (3*pi/8)*tmp*tmp*tmp;
    bound = exp(-x*x/(2*sigma*sigma))/(sigma*sqrt(2*pi));
    test = rand()*c*bound;
    if (test < y)
      picking = 0;
      samples(1,i) = x;
      samples(2,i) = y;
    end
  end
  Ex = Ex + x;
  Ex2 = Ex2 + x*x;
end

Ex = Ex/N
Ex2 = Ex2/N

scatter(samples(1,:),samples(2,:));
